src/Workspaces/CSharp/Portable/Utilities/CSharpUseExplicitTypeHelper.cs

```diff
@@ -96,9 +96,9 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
                     return false;
                 }
             }
-            else if (typeName.Parent.IsKind(SyntaxKind.ForEachStatement))
+            else if (typeName.Parent is ForEachStatementSyntax foreachStatement &&
+                     foreachStatement.Type == typeName)
```

Comment:
can you add a test for this too? it seems like the test would only cover the other change

Reply:
That's harder to do.  The UseExplicitTypeHelper instance is only used from one location (UseExplicitTypeCodeRefactoringProvider) and that type already first ensures it is only operating on the 'type' node of the foreach, through CSharpTyepStyleHelper.FindAnalyzableType.

However, i wanted to be very paranoid here so that if this moves around in the future, we don't *accidently* starting lighting this up.  (i.e. the same sort of issue we ran into when i moved things around originally, and now that exact problem happened).

Note: i'm happy to add a pseudo-test for this. i.e. one where this is "var" in the collection part of a foreach.  however, it will act more as a guard in the future, and not actually reflect a fix in any broke behavior today.

Reply:
Added paranoia tests to several of the test harnesses that uses these helpers.

Reply:
> The UseExplicitTypeHelper instance is only used from one location (UseExplicitTypeCodeRefactoringProvider) and that type already first ensures it is only operating on the 'type' node of the foreach

my question would then be: why does this code exist in two places, as opposed having one check for it here? otherwise there could be subtle differences or bugs when this is used from another place

Reply:
Effectively, this code has had a long and tumultuous history.  We've had code for this feature scattered in many different locations.  My original PRs were an attempt to address this precisely because you coudl observe different behavoir depending on what codepaths you went through.  I'm absolutely sure there are more opportunities for unification for the var/explicit-type code.   :)

However, i thought it woudl be best to try to keep the code changes here targeted and minimal to reduce the chance of risk/regression.  I think it would be fine for later PRs to come along with further improvements here to make it less easy to get into a bad state.

I'll leave it to other PR reviewers to suggest if they think more unification should be done here, or if hte fix should be as is.

Reply:
Ok, it wasn't my intention to say you should do the refactoring here. Thanks.

Reply:
all if statements above this also just check parent kind, but doesn't check whether typename is on right children side. can others cause same issue?

Reply:
Yes.  I could see some issues here.  Let me try to examine some other potential concerning locations.

Reply:
Ok.  so looking at UseImplicitTypeHelper.  We do the following checks:

1a. is our parent a VariableDeclaration.  In that case, it could only be the "type".  so all checks beyond that are ok.
2a. is our parent a foreach. i just fixed this case.
3a. is our parent a declaration expression.  in that case, it could only be hte "type" portion.

For UseExplicitTypeHelper we do the following checks:

1b. Is this a declaration expression.  Like 3a, this is ok.
2b. Is this a variable declaration.  Like 1a this is ok.
3b. Is this a foreach.  This is waht i hardened.

So, *for now*, we're ok.  But we could certainly make 1a, 3a, 1b and 2b tighter if we wanted.  I don't think they're necessary for this PR though.

I'm reasonably confident we don't have further regressions *in these particular methods*.  But i can't discount other codepaths havign problems as there might be something else i missed elsewhere :-/